consensus/src/processes/relations.rs

```diff
@@ -0,0 +1,159 @@
+use super::ghostdag::mergeset::unordered_mergeset_without_selected_parent;
+use crate::model::{services::reachability::ReachabilityService, stores::relations::RelationsStore};
+use itertools::Itertools;
+use kaspa_consensus_core::{
+    blockhash::{BlockHashIteratorExtensions, BlockHashes, ORIGIN},
+    BlockHashSet,
+};
+use kaspa_database::prelude::{BatchDbWriter, DbWriter, StoreError};
+use kaspa_hashes::Hash;
+use rocksdb::WriteBatch;
+
+/// Initializes this relations store with an `origin` root
+pub fn init<S: RelationsStore + ?Sized>(relations: &mut S) {
+    if !relations.has(ORIGIN).unwrap() {
+        relations.insert(ORIGIN, BlockHashes::new(vec![])).unwrap();
+    }
+}
+
+/// Delete relations of `hash` for the case where the relations store represents a specific level.
+/// In this case we simply remove the entry locally, relying on the fact that level relations are
+/// kept topologically continuous. If any child of this `hash` will remain with no parent, we make
+/// sure to connect it to `origin`. Note that apart from the special case of `origin`, these relations
+/// are always a subset of the original header relations for this level.
+pub fn delete_level_relations<W, S>(mut writer: W, relations: &mut S, hash: Hash) -> Result<(), StoreError>
+where
+    W: DbWriter,
+    S: RelationsStore + ?Sized,
+{
+    let children = relations.get_children(hash)?; // if the first entry was found, we expect all others as well, hence we unwrap below
+    for child in children.iter().copied() {
+        let child_parents = relations.get_parents(child).unwrap();
+        // If the removed hash is the only parent of child, then replace it with `origin`
+        let replace_with: &[Hash] = if child_parents.as_slice() == [hash] { &[ORIGIN] } else { &[] };
+        relations.replace_parent(&mut writer, child, hash, replace_with).unwrap();
+    }
+    relations.delete(&mut writer, hash).unwrap();
+    Ok(())
+}
+
+/// Delete relations of `hash` for the case where relations represent the maximally known reachability
+/// relations. In this case we preserve all topological info by connecting parents of `hash` as parents
+/// of its children if necessary. This means that these relations do not correlate with header data and
+/// can contain links which didn't appear in the original DAG (but yet follow from it).
+///
+/// NOTE: this algorithm does not support a batch writer bcs it might write to the same entry multiple times
```

Comment:
yes, if you write batch once and the entry you modified got out of the cache, then the next time you try updating it, it will be read from the DB without the first update